### ISSUE TITLE
Provide the monadic "join" function

### ISSUE THREAD

**Issue by nikita-volkov** [Thomvis/BrightFutures#41]
I'm talking about something like the following:
```swift
func join<A>(future: Future<Future<A>>) -> Future<A> {
  return future.flatMap{$0}
}
```

**Comment by Thomvis** [comment]
Thanks for your input!

Three points for discussion:
- This would be great to have if Swift would support extending generic classes that have a specific type signature (e.g. a Future<T> where T is Future<U>). Since this is not the case, `join` will have to be a public global/top-level function, which I'm not very comfortable with, but that might be my OOP background speaking.
- Does it really add substantial convenience/clarity over `future.flatMap{$0}`
- I understand that `join` is the terminology used in Haskell, but for me it would make more sense to name this function `flatten`. Does that make sense?

**Comment by nikita-volkov** [comment]
You're correct about the Haskell reference. However, "Future" forms a monad and ["join" is the standard name of such an operation](http://en.wikipedia.org/wiki/Monad_(functional_programming)#fmap_and_join) in the context of a monad. People with functional background will be looking for it and they will be looking for it under that exact name. I know that, because that's what I did myself.

So I vote for "join". If Scala's standard library didn't mess people's minds up with new definitions, I'd also ask for the proper "bind" instead of "flatMap" as well, but I sense you wouldn't be inclined to changing that.

BTW, instead of `future.flatMap{$0}` it should probably be implemented more efficiently by directly using the internal mechanisms of the library.

**Comment by Thomvis** [comment]
I appreciate the elegance of functional programming concepts and BrightFutures is definitely inspired by them in a significant way. The usability of the library for people coming from a functional background does matter to me. You'll find that BrightFutures is more functional than some other Future implementations in Swift out there.

That said, the primary goal for BrightFutures is to provide idiomatic futures & promises for Swift. It should work with, and act like, the Swift standard library as closely as possible. Swift 1.2 introduced `flatMap`, so that name is a given. The question now is given `flatMap`, what name makes the most sense in Swift for the join/flatten operation?

**Comment by nikita-volkov** [comment]
I agree, in the context of `flatMap` the name `flatten` does seem consecutive.

**Comment by nikita-volkov** [comment]
One option is to provide references to the standard names of those functions in the docs.

**Comment by Thomvis** [comment]
That's a great idea. If you want, feel free to start a PR on either of the two tasks. If not, I'll do it, but perhaps after 1.0.

Thanks for understanding.

**Comment by mdedetrich** [comment]
Isn't BrightFuture's meant to be inspired by the Scala Future API, in which case, `map` and `flatMap` do what they are supposed to do?


PS. arguing about what the correct terminology is kind of redundant, The names `map` and `flatMap` have no real definition, the actual definition are mathematical statements. Haskell just happened to use words such as `bind` because thats what they thought was intuitive.

Honestly this is just an argument about whether we should be using Haskell terminology or Scala terminology

**Comment by nikita-volkov** [comment]
> Haskell just happened to use words such as bind because thats what they thought was intuitive.

Haskell doesn't use `bind`. It uses `>>=`.

The name "bind" is part of the universal terminology used in the functional world, which is not limited to Haskell and Scala. Amongst prominent players there also exist F#, OCaml, Erlang, Clojure and even JavaScript.

> Honestly this is just an argument about whether we should be using Haskell terminology or Scala terminology

Scala's standard library doesn't even mention monads. The "Scalaz" library is what brings the functional abstractions to the language, and FYI [it uses "bind"](https://github.com/scalaz/scalaz/blob/c847654dcf75c748eacbaf246511bbd938b8631f/core/src/main/scala/scalaz/Bind.scala#L16).

**Comment by mdedetrich** [comment]
I think you are missing the point I was working, Scala is using different terminology (and it just that, different terminology), and the the BrightFuture's library was based off the scala API. Its even mentioned in the blurb

```The goal of this project is to port Scala's Futures & Promises (guide, api) to Swift. Second to this readme, the Scala documentation should therefore also be of help.```

Furthermore, in Swift 1.2 (just released yesterday), they also use `flatMap` the same way Scala does, i.e. http://airspeedvelocity.net/2015/03/13/changes-to-the-swift-standard-library-in-1-2-betas-2-and-3/

In other words, in terms of consistency with both the goal of the project, and with the Swift API, there isn't much of a reason to change `flatMap`.

**Comment by Thomvis** [comment]
> and the the BrightFuture's library was based off the scala API. Its even mentioned in the blurb

I have to admit that I might want to revise that goal. BrightFutures started out as a port of Scala's Future, but its current goal can be better worded (as I did above) as "to be an idiomatic futures library for Swift". So far, these two goals have been aligned, including the naming of the `flatMap` operation.

I think we all agree that the naming of `flatMap` (and `flatten`) are probably best for Swift and thus for BrightFutures. Thanks for your input!

**Comment by nikita-volkov** [comment]
> I think we all agree that the naming of flatMap (and flatten) are probably best for Swift and thus for BrightFutures.

Yeah. I'm not arguing that.

**Comment by durre** [comment]
Even if it's not a goal anymore, coming from Scala (as I am), I feel right at home with BrightFutures :)
